### PR TITLE
Feature/peaks

### PR DESCRIPTION
**Pull request by tremblap**
this PR is proposing a new descriptor, extracting sinusoidal peaks from the signal. it has a similar interface to all 'feature' descriptors, to help understand the underlying algorithm for more complex objects, in this case, the (buf)sines one. The voice allocation will be done in another object.

### CODE REVIEW

File: include/algorithms/public/SineFeatureExtraction.hpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+#include "../../data/TensorTypes.hpp"
+#include <Eigen/Core>
+#include <cmath>
+#include <queue>
+
+namespace fluid {
+namespace algorithm {
+
+
+class SineFeatureExtraction
```
Comment (weefuzzy) [inline include/algorithms/public/SineFeatureExtraction.hpp:30]:
```suggestion
class SineFeature
```

File: include/algorithms/public/SineFeatureExtraction.hpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+class SineFeatureExtraction 
+{
+  using ArrayXd = Eigen::ArrayXd;
+  using VectorXd = Eigen::VectorXd;
+  using ArrayXcd = Eigen::ArrayXcd;
+  template <typename T>
+  using vector = rt::vector<T>;
+
+public:
+  SineFeatureExtraction(Allocator& alloc)
```
Comment (weefuzzy) [inline include/algorithms/public/SineFeatureExtraction.hpp:39]:
```suggestion
SineFeature(Allocator& alloc)
```

Comment (weefuzzy) [reply include/algorithms/public/SineFeatureExtraction.hpp:39]: We will, at some point, want to pass maximal window / fftSizes in by the constructor too

File: include/algorithms/public/SineFeatureExtraction.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  {
+    mBins = fftSize / 2 + 1;
+    mScale = 1.0 / (windowSize / 4.0); // scale to original amplitude
+    mInitialized = true;
+  }
+
+  void processFrame(const ComplexVectorView in, RealVectorView freqOut,
+                    RealVectorView magOut, index logFreq, index logMag,
+                    double sampleRate, double detectionThreshold,
+                    index sortBy, Allocator& alloc)
```
Comment (weefuzzy) [inline include/algorithms/public/SineFeatureExtraction.hpp:52]: `index sortBy` can be `bool sort` again

File: include/algorithms/public/SineFeatureExtraction.hpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      freqOut(Slice(maxNumOut, freqOut.size() - maxNumOut)).fill(0);//pad the size with "no-pitch" (0Hz);      
+    }
+    
+    if (logMag) {
+      std::transform(tmpPeaks.begin(),tmpPeaks.begin()+maxNumOut,magOut.begin(),[](auto peak){return peak.second;});
+      magOut(Slice(maxNumOut, freqOut.size() - maxNumOut)).fill(-144);//pad the size with 'silence' in dB;
+    } else {
+      std::transform(tmpPeaks.begin(),tmpPeaks.begin()+maxNumOut,magOut.begin(),[](auto peak){return std::pow(10, (peak.second / 20));});
+      magOut(Slice(maxNumOut, freqOut.size() - maxNumOut)).fill(0);//pad the size with silence;    
+    }
```
Comment (weefuzzy) [inline include/algorithms/public/SineFeatureExtraction.hpp:86]: All this belongs in the *client* to be consistent with similar objects

File: include/algorithms/util/PeakDetection.hpp [hunk 5]
```diff
@@ -29,7 +29,7 @@ class PeakDetection
 public:
   pairs_vector process(const Eigen::Ref<ArrayXd>& input, index numPeaks = 0,
                        double minHeight = 0, bool interpolate = true,
-                       bool       sort = true,
+                       bool sortByMag = false,
```
Comment (weefuzzy) [inline include/algorithms/util/PeakDetection.hpp:32]: You can remove this change

File: include/algorithms/util/PeakDetection.hpp [hunk 6]
```diff
@@ -56,11 +56,11 @@ class PeakDetection
...
       }
     }
-    if (sort)
+    if (sortByMag)
     {
-      std::sort(peaks.begin(), peaks.end(), [](auto& left, auto& right) {
+        std::sort(peaks.begin(), peaks.end(), [](auto& left, auto& right) {
         return left.second > right.second;
-      });
+        });
```
Comment (weefuzzy) [inline include/algorithms/util/PeakDetection.hpp:63]: you can revert these changes as well, and `PeakDetection` no longer has any changes in this PR

File: include/clients/common/FluidNRTClientWrapper.hpp [hunk 7]
```diff
@@ -637,7 +637,7 @@ struct StreamingControl
     for (auto outs = std::pair{outputBuffers.begin(), outputData.begin()};
          outs.first != outputBuffers.end(); ++outs.first, ++outs.second)
     {
-      if (!outs.first) continue;
+      if (!(*outs.first)) continue;
```
Comment (weefuzzy) [inline include/clients/common/FluidNRTClientWrapper.hpp:640]: This change is a bugfix for `dev`, not this PR. Please apply directly to `dev`

File: include/clients/rt/SineFeatureClient.hpp [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+#include "../common/AudioClient.hpp"
+#include "../common/BufferedProcess.hpp"
+#include "../common/FluidBaseClient.hpp"
+#include "../common/FluidNRTClientWrapper.hpp"
+#include "../common/ParameterConstraints.hpp"
+#include "../common/ParameterSet.hpp"
+#include "../common/ParameterTrackChanges.hpp"
+#include "../common/ParameterTypes.hpp"
+#include "../../algorithms/public/SineFeatureExtraction.hpp"
```
Comment (weefuzzy) [inline include/clients/rt/SineFeatureClient.hpp:20]:
```suggestion
#include "../../algorithms/public/SineFeature.hpp"
```

File: include/clients/rt/SineFeatureClient.hpp [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  kFFT
+};
+
+constexpr auto SineFeatureParams = defineParameters(
+    LongParamRuntimeMax<Primary>("numPeaks", "Number of Sinusoidal Peaks", 10,
+                        Min(1),
+                        FrameSizeUpperLimit<kFFT>()),
+    FloatParam("detectionThreshold", "Peak Detection Threshold", -96, Min(-144),
+               Max(0)),
+    EnumParam("sortBy", "Sort Peaks Output", 0, "Frequencies", "Amplitudes"),
```
Comment (weefuzzy) [inline include/clients/rt/SineFeatureClient.hpp:42]:
```suggestion
EnumParam("order", "Sort Peaks Output", 0, "Frequencies", "Amplitudes"),
```

File: include/clients/rt/SineFeatureClient.hpp [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  kLogFreq,
+  kLogMag,
+  kFFT
+};
+
+constexpr auto SineFeatureParams = defineParameters(
+    LongParamRuntimeMax<Primary>("numPeaks", "Number of Sinusoidal Peaks", 10,
+                        Min(1),
+                        FrameSizeUpperLimit<kFFT>()),
+    FloatParam("detectionThreshold", "Peak Detection Threshold", -96, Min(-144),
```
Comment (weefuzzy) [inline include/clients/rt/SineFeatureClient.hpp:40]: what were we smoking when we decided `detectionThreshold` was a decent name in `SinesClient`? Guess we're stuck with it now

File: include/clients/rt/SineFeatureClient.hpp [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        
+    output[0](Slice(0, nPeaks)) <<= peaks;
+    if (get<kLogFreq>()) {
+      output[0](Slice(nPeaks, get<kNPeaks>().max() - nPeaks)).fill(-999);
+    } else {
+      output[0](Slice(nPeaks, get<kNPeaks>().max() - nPeaks)).fill(0);      
+    }
+
+    output[1](Slice(0, nPeaks)) <<= mags;
+    if (get<kLogFreq>()) {
```
Comment (weefuzzy) [inline include/clients/rt/SineFeatureClient.hpp:121]:
```suggestion
if (get<kLogMag>()) {
```

**Review by weefuzzy — CHANGES_REQUESTED** [review]
So the headlines: 
* `PeakDetection.hpp` no longer needs any changes to it, so can be reset
* changes to `FluidNRTClientWrapper` should be applied on `dev` with an on-topic commit msg
* rename `SineFeatureExtraction.hpp` -> `SineFeature.hpp` for consistency 

Overall things: 
* Need to run `clang-format` on new and changed files before merge, can do it or talk you through it 
* Need to review whether the duplicated code between `SineFeature` and `SineExtraction` is egregious enough to change the latter to call through the former.

File: include/algorithms/util/PeakDetection.hpp (hunk 5 above)
Comment (tremblap) [reply include/algorithms/util/PeakDetection.hpp:32]: i thought it made the code clearer...

File: include/clients/common/FluidNRTClientWrapper.hpp (hunk 7 above)
Comment (tremblap) [reply include/clients/common/FluidNRTClientWrapper.hpp:640]: ok

File: include/clients/rt/SineFeatureClient.hpp (hunk 11 above)
Comment (tremblap) [reply include/clients/rt/SineFeatureClient.hpp:121]: nice catch!

File: include/clients/common/FluidNRTClientWrapper.hpp (hunk 7 above)
Comment (tremblap) [reply include/clients/common/FluidNRTClientWrapper.hpp]: done

File: include/algorithms/public/SineFeatureExtraction.hpp (hunk 2 above)
Comment (tremblap) [reply include/algorithms/public/SineFeatureExtraction.hpp]: for now, done.

File: include/algorithms/public/SineFeatureExtraction.hpp (hunk 4 above)
Comment (tremblap) [reply include/algorithms/public/SineFeatureExtraction.hpp]: ok I remember now why I did this in the algo. we need to fill a discrete value depending on how many peaks are actually returned by PeakDetection (depending on thresholding). if we are in exponential mode the value is different.

so plan A: we leave as is.
plan B: we fill in 2 places - with the default value in the algo (0) then with a search within the client for that value to be replaced if log... we could check from the end of the returned array until the value is no more 0...

Comment (weefuzzy) [reply include/algorithms/public/SineFeatureExtraction.hpp]: Just have the algorithm return the number of detected peaks from `processFrame`.

Comment (tremblap) [reply include/algorithms/public/SineFeatureExtraction.hpp]: I thought of it and found it dirty. will do.

File: include/clients/rt/SineFeatureClient.hpp [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+              in.row(0), peaks, mags, sampleRate(), get<kDetectionThreshold>(),
+              get<kSortBy>(), c.allocator());
+        });
+
+    auto validPeaks = mPeaks(Slice(0, mNumPeaks));
+    auto validMags = mMags(Slice(0, mNumPeaks));
+
+    if (get<kLogFreq>())
+    {
+      auto ratio = 1 / 440.0;
```
Comment (weefuzzy) [inline include/clients/rt/SineFeatureClient.hpp:118]: if this never ever changes, we can make it constexpr
